### PR TITLE
Make block generator easier to use

### PR DESCRIPTION
**Pull request by debris**
let's make it easier to write tests for blockchain

### CODE REVIEW

File: ethcore/src/blockchain/blockchain.rs [hunk 1]
```diff
@@ -1524,101 +1522,99 @@ mod tests {
...
-		let b2a = canon_chain.generate(&mut finalizer).unwrap();
-		let b3b = canon_chain.fork(1).generate(&mut finalizer.fork()).unwrap();
-		let b3a = canon_chain.generate(&mut finalizer).unwrap();
-		let b4b = canon_chain.fork(1).generate(&mut finalizer.fork()).unwrap();
-		let b4a = canon_chain.generate(&mut finalizer).unwrap();
-		let b5b = canon_chain.fork(1).generate(&mut finalizer.fork()).unwrap();
-		let b5a = canon_chain.generate(&mut finalizer).unwrap();
+		let genesis = BlockBuilder::genesis();
+		let b1a = genesis.add_block();
+		let b2a = b1a.add_block();
```
Comment (tomusdrw) [inline ethcore/src/blockchain/blockchain.rs]: Why not:
```
let chain_a = genesis.add_blocks(5);
let chain_b = genesis.add_blocks_with_difficulty(9);

let uncle_headers = chain_b[2..4].rev().map(|x| x.last_block_header()).collect();
```
or
```
let chain_a = genesis.add_blocks(5);
// with_difficulty clones the builder and changes the default options for all new generated blocks
let chain_b = genesis.with_difficulty(9).add_blocks(5);
```

File: ethcore/src/blockchain/blockchain.rs [hunk 2]
```diff
@@ -1501,18 +1500,17 @@ mod tests {
 	#[test]
 	fn should_cache_best_block() {
 		// given
-		let mut canon_chain = ChainGenerator::default();
-		let mut finalizer = BlockFinalizer::default();
-		let genesis = canon_chain.generate(&mut finalizer).unwrap();
-		let first = canon_chain.generate(&mut finalizer).unwrap();
+		let genesis = BlockBuilder::genesis();
+		let first = genesis.add_block();
+		let mut generator = BlockGenerator::new(vec![genesis, first]);
```
Comment (tomusdrw) [inline ethcore/src/blockchain/blockchain.rs]: IMHO `BlockGenerator` should not be `Iterator`, but rather `IntoIterator` (Maybe even call it `BlockChain`)
```
let generator = BlockGenerator::new(&[genesis, first]);
new_chain(&generator[0].data())
// Retrieve iterator
let mut it = generator.iter();

// or consume generator and do sth with blocks
for b in generator {
  ...
}

File: ethcore/src/blockchain/generator.rs [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+impl Block {
+	#[inline]
+	pub fn hash(&self) -> H256 {
+		BlockView::new(&self.data()).header_view().hash()
+	}
+
+	#[inline]
+	pub fn data(&self) -> Bytes {
```
Comment (tomusdrw) [inline ethcore/src/blockchain/generator.rs]: `data` is imho too generic. Let's just call it `encoded()` maybe?

File: ethcore/src/blockchain/generator.rs [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+		encode(self).into_vec()
+	}
+}
+
+#[derive(Debug)]
+pub struct BlockMetadata {
```
Comment (tomusdrw) [inline ethcore/src/blockchain/generator.rs]: Would rather call it `BlockOptions`

File: ethcore/src/blockchain/generator.rs [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+			blocks.push_back(block);
+		}
+
+		BlockBuilder {
+			blocks,
+		}
+	}
+
+	#[inline]
+	pub fn last_block_hash(&self) -> H256 {
```
Comment (tomusdrw) [inline ethcore/src/blockchain/generator.rs]: Imho `block` is redundant, maybe do this:
```
blocks.last().number()
blocks.last().hash()
blocks.last().header()
blocks.last().into::<Block>()
```

Comment (debris) [reply ethcore/src/blockchain/generator.rs]: 👍 , if `blocks.last()` is already a `Block`, there is no need for into :)

File: ethcore/src/blockchain/blockchain.rs (hunk 1 above)
Comment (debris) [reply ethcore/src/blockchain/blockchain.rs]: if you look closely, there are 5 different forks :)

```
genesis
|      \ 
b1a     b1b
|   \
|    \
|     \
|      \
b2a     b2b
|
....
```

Comment (debris) [reply ethcore/src/blockchain/blockchain.rs]: I didn't want to change the logic of a test

**Comment by svyatonik** [comment]
Test failure was fixed in master => merged master here

**Comment by debris** [comment]
I removed usages of `generator.next().unwrap()` :)

**Pull request merged**